### PR TITLE
ENH: Add MSTA Bits

### PR DESCRIPTION
**Pull request by slactjohnson**
## Description
Adds in an EPICS signal to bring the the motor MSTA bits. This signal is used to create a dictionary of interpreted values, as well as a "homed" property. 

## Motivation and Context
I would like to implement some automated homing on some finicky motor stacks. I need a way to check that the homing routine has completed.  

## How Has This Been Tested?
Tested live on one of the stages of interest (a SmarAct linear stage). 

## Where Has This Been Documented?
This PR. 

## Screenshots:
Live session (standard motor record):
image

IMS Motor:

happy_ims

sad_ims

sad_ims_2

image

Newport Motor:

happy_newport

sad_newport

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

File: pcdsdevices/epics_motor.py
Comment (tangkong) [inline pcdsdevices/epics_motor.py:52]: gotta love EPICS.  Is this a bitmask thing?

Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:52]: Yep. The field and bit descriptions are in a table here: https://epics.anl.gov/bcda/synApps/motor/R5-6/motorRecord.html#Fields_status

Comment (tangkong) [reply pcdsdevices/epics_motor.py:52]: That doc says bit 6 is also unused.  Is our motor record different?

Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:52]: My bad. That's an old version of the documentation. Here's the latest, which I believe matches what I have here: https://epics.anl.gov/bcda/synApps/motor/motorRecord.html#Fields_status

File: pcdsdevices/epics_motor.py
Comment (tangkong) [inline pcdsdevices/epics_motor.py:237]:
```python
d = {bit.name: (val >> bit.value) & 0x1 for bit in MstaEnum}  
```
A fun pythonic dict-comprehension could make this a one-liner.

A more meaningful comment: I wonder if there's a clearer way to do these bit shifts.  It took my python mind a while to understand the simple bit-extraction you're doing here.  

Maybe `enum.IntFlag`?
```python
from enum import IntFlag, auto

class MstaEnum(IntFlag):
    """
    Enum for the EPICS motor record .MSTA field bits.
    """
    # Note: the motor record docs start bit numbering at 1, but the MSTA bits
    # start at 0.
    direction = auto()       # last raw move direction (0: negative, 1: positive)
    done = auto()            # motion is complete
    plus_ls = auto()         # plus limit switch is hit
    home_ls = auto()         # state of the home limit switch
    # Bit 4 is un-used
    closed_loop = auto()
    slip_stall = auto()      # slip stall is detected
    home = auto()            # at the home position
    enc_present = auto()     # encoder is present. Called "present" in the docs.
    problem = auto()         # driver stopped polling, or there's a hardware problem
    moving = auto()         # the motor has a non-zero velocity (it's moving)
    gain_support = auto()   # the motor supports closed loop control
    comm_error = auto()     # controller communication error
    minus_ls = auto()       # minus limit switch is hit
    homed = auto()          # the motor has been homed


print(MstaEnum(0))  # >> MstaEnum.0
print(MstaEnum(129))  # >> MstaEnum.enc_present|direction

print([flag.name for flag in MstaEnum if flag in MstaEnum(129)])  # >> ['direction', 'enc_present]
```

Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:237]: That's a good one-liner. I forgot about dictionary comprehensions. Sometimes my python ends up a bit C-ish. :)

Does `auto()` work here? I'm not familiar with it but from the docs it sounds like it just generates sequential numbers. The raw MSTA value that we're getting from the IOC is based on the mapping in the motor record, so the values matter, and I think we'd get off by one after the missing 4th bit.

File: pcdsdevices/epics_motor.py
Comment (silkenelson) [reply pcdsdevices/epics_motor.py:52]: Please make sure this work as expected. I seem to recall ( but am not 100% positive) that our very special IMS2 record as redefined the MSTA fields.

Comment (silkenelson) [reply pcdsdevices/epics_motor.py:52]: Similarly, I'd suggest to also take a look at the Newport motor record.

Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:52]: @silkenelson You're right. The IMS record and the Newport record does redefine some of these bits. I didn't see anything odd in the MCS2 motor record. I'll look into how to deal with this on Monday. 

Vanilla Motor Record: 

image

Newport Motor Record: 

image

Latest IMS motor record:

Screenshot from 2024-07-21 09-34-46

File: pcdsdevices/epics_motor.py
Comment (ZLLentz) [reply pcdsdevices/epics_motor.py:237]: I think to use auto (and an int flag in general) you'd have to include a placeholder bit

File: pcdsdevices/epics_motor.py
Comment (tangkong) [reply pcdsdevices/epics_motor.py:52]: I think the most natural way to deal with this is to make different `MstaEnum`'s for different motor records.  `EpicsMotorInterface` might use the default one, then `pcdsdevices.epics_motor.Newport` gets its own, etc.

This is assuming we motor subclasses for these records.

Comment (tangkong) [reply pcdsdevices/epics_motor.py:52]: It looks like we have the following `EpicsMotorInterface` subclasses to consider from the `pcdsdevices` side of things.  (Of course we could have motors that don't have their own subclass)

A quick (turned not-so-quick) survey of the `EpicsMotorInterface` family tree

* `pcdsdevices.ccm.CCMPico`
* `pcdsdevices.epics_motor.PCDSMotorBase`
  * `pcdsdevices.epics_motor.IMS`
  * `pcdsdevices.epics_motor.Newport`
  * `pcdsdevices.epics_motor.DelayNewport` 
    * `pcdsdevices.lxe.TimeToolDelay`
    * `pcdsdevices.lxe._ReversedTimeToolDelay`
  * `pcdsdevices.epics_motor.PMC100`
  * `pcdsdevices.epics_motor.MMC100`
* `pcdsdevices.epics_motor.BeckhoffAxis`
  * `pcdsdevices.epics_motor.BechoffAxisEPS`
  * `pcdsdevices.epics_motor.BechoffAxisEPSCustom`
  * `pcdsdevices.epics_motor.BechoffAxis_Pre140` 
  * `pcdsdevices.epics_motor.BechoffAxisNoOffset`
* `pcdsdevices.epics_motor.SmarAct`


An interesting mapping we have:
```
    motor_types = (('MMS', IMS),
                   ('CLZ', IMS),
                   ('CLF', IMS),
                   ('MMN', Newport),
                   ('MZM', PMC100),
                   ('MMC', MMC100),
                   ('MMB', BeckhoffAxis),
                   ('PIC', PCDSMotorBase),
                   ('MCS', SmarAct),
                   ('MCS2', SmarAct))
```

Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:52]: > I think the most natural way to deal with this is to make different `MstaEnum`'s for different motor records. `EpicsMotorInterface` might use the default one, then `pcdsdevices.epics_motor.Newport` gets its own, etc.
> 
> This is assuming we motor subclasses for these records.

I agree. This is the direction I've taken in my dev branch, just waiting to test.

File: pcdsdevices/epics_motor.py
Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:237]: Now that I see the IMS and Newport IOCs have members that are larger than one bit, I think `auto()` will not work, or at least not easily or cleanly. I think I'd rather stick with an explicit mapping rather than implicitly making it with `auto()`.

Comment (tangkong) [reply pcdsdevices/epics_motor.py:237]: I think `IntFlag` still works if you define the bits manually, but I agree `auto()` won't quite work here.
```python
class MstaEnum(IntFlag):
    """
    Enum for the EPICS motor record .MSTA field bits.
    """
    # Note: the motor record docs start bit numbering at 1, but the MSTA bits
    # start at 0.
    direction = 1       # last raw move direction (0: negative, 1: positive)
    done = 2            # motion is complete
    plus_ls = 4        # plus limit switch is hit
    home_ls = 8         #
    ...
```

**Comment by slactjohnson** [comment]
@silkenelson I've now set up separate enums and decoding for the IMS and Newport motors. It appears to work as expected (see the images in the PR description). 

Unfortunately, I wasn't able to find a Newport or IMS motor that was reporting an actual error _code_. Is there anything you can think of that would force an error code on one of these stages? Is there anything else you would like to see in terms of verifying functionality?

**Comment by ZLLentz** [comment]
> Is there anything else you would like to see in terms of verifying functionality?

I think the screenshots in your description are sufficient for functionality

I'm somewhat curious about the other various motor types in the tree and if they build off of the standard MSTA or off of one more like IMS/Newport. I'm eternally frustrated by these "special" record types, this seems to bite us surprisingly frequently.

File: docs/source/upcoming_release_notes/1205-Add_MSTA_field_and_homed_status_to_EpicsMotorInterface.rst
Comment (ZLLentz) [inline docs/source/upcoming_release_notes/1205-Add_MSTA_field_and_homed_status_to_EpicsMotorInterface.rst:16]: I think the only thing missing is including that you also did this separately for IMS and Newport

File: pcdsdevices/epics_motor.py
Comment (ZLLentz) [inline pcdsdevices/epics_motor.py:294]: This is so incredibly puzzling to me. The manual says that it's supposed to be a ULONG but it's clearly a DOUBLE if you check any of the fields...

**Review by ZLLentz — APPROVED** [review]
I'm full :+1: on this implementation

File: pcdsdevices/epics_motor.py
Comment (tangkong) [reply pcdsdevices/epics_motor.py:294]: This isn't something we did specific to our records right?...

**Review by tangkong — APPROVED** [review]
Looks good to me, once again we document EPICS weirdness better in python

File: pcdsdevices/epics_motor.py
Comment (ZLLentz) [reply pcdsdevices/epics_motor.py:294]: It seems to be universal between standard records and the weird stuff we created here. I can't rule out a strange custom motor record build I guess.

**Comment by ZLLentz** [comment]
I'm going to click "Update branch" so hopefully it's ready to merge when we look again tomorrow

**Comment by ZLLentz** [comment]
CI failed on some network issue- running again before I sign off

File: pcdsdevices/epics_motor.py
Comment (slactjohnson) [reply pcdsdevices/epics_motor.py:294]: Looking at our latest motor record (R6.10.1-2.3.3), it's a DBF_ULONG in motorRecord.dbd as well as in e.g. ims.dbd, so I'm not sure why it's a float. 
image
image

**Comment by slactjohnson** [comment]
Going to wait on merging for a little bit for @silkenelson to comment on the implementation or any other tests that should be done.

**Comment by slactjohnson** [comment]
I'm going to go ahead and merge this because some of my current work is waiting on this update. We can revise later if needed.

**Pull request merged**
